camel-core/src/main/java/org/apache/camel/component/saga/SagaProducer.java

```diff
@@ -0,0 +1,78 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.component.saga;
+
+import org.apache.camel.AsyncCallback;
+import org.apache.camel.Exchange;
+import org.apache.camel.impl.DefaultAsyncProducer;
+import org.apache.camel.saga.CamelSagaService;
+import org.apache.camel.util.CamelContextHelper;
+
+/**
+ * A producer that finalizes the current saga.
+ */
+public class SagaProducer extends DefaultAsyncProducer {
+
+    private final boolean success;
+
+    private CamelSagaService camelSagaService;
+
+    public SagaProducer(SagaEndpoint endpoint, boolean success) {
+        super(endpoint);
+        this.success = success;
+
+        CamelSagaService sagaService = endpoint.getCamelContext().hasService(CamelSagaService.class);
+        if (sagaService == null) {
+            sagaService = CamelContextHelper.findByType(endpoint.getCamelContext(), CamelSagaService.class);
+        }
+        if (sagaService == null) {
+            throw new IllegalStateException("Cannot find saga service: saga producers can only be used within a saga");
+        }
+        this.camelSagaService = sagaService;
+    }
+
+    @Override
+    public boolean process(Exchange exchange, AsyncCallback callback) {
+        String currentSaga = exchange.getIn().getHeader(Exchange.SAGA_LONG_RUNNING_ACTION, String.class);
+        if (currentSaga == null) {
+            exchange.setException(new IllegalStateException("Current exchange is not bound to a saga context: cannot complete"));
+            callback.done(true);
+            return true;
+        }
+
+
+        camelSagaService.getSaga(currentSaga).thenApply(coordinator -> {
+            if (coordinator == null) {
+                throw new IllegalStateException("No coordinator found for saga id " + currentSaga);
```

Comment:
Should you not set this exception on exchange and call callback done false, to ensure camel routing engine continues properly

Reply:
That's done on the `whenComplete` callback few lines below.

Reply:
Ah okay so it ensures all thrown exception is handled by it and it will always call the whenComplete callback.